### PR TITLE
Add errorfile stanzas and dummy default html files to the router

### PR DESCRIPTION
**Pull request by miheer**
Add errorfile stanzas and dummy default html files to the router
https://issues.redhat.com/browse/NE-379

### CODE REVIEW

**Comment by miheer** [comment]
@Miciah PTAL

File: images/router/haproxy/conf/error-page-404.http [hunk 1]
```diff
@@ -0,0 +1,148 @@
+HTTP/1.0  404 Error, Page Not Found, File Not Found, or Server Not Found
```
Comment (Miciah) [inline images/router/haproxy/conf/error-page-404.http:1]: I'm not sure what the restrictions on the reason phrase are, but reason phrases are usually much more terse:
```suggestion
HTTP/1.0 404 Not Found
```

File: images/router/haproxy/conf/error-page-404.http [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    }
+    h1 {
+      font-size: 2.15em;}
+  }
+
+  </style>
+  </head>
+  <body>
+    <div>
+      <h1>404 Error, Page Not Found, File Not Found, or Server Not Found</h1>
```
Comment (Miciah) [inline images/router/haproxy/conf/error-page-404.http:116]: Let's pick just one:
```suggestion
      <h1>Application or Document Not Found</h1>

```

File: images/router/haproxy/conf/error-page-404.http [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+      <p>The browser/client was able to communicate with a given server, but the server could not find what was requested</p>
```
Comment (Miciah) [inline images/router/haproxy/conf/error-page-404.http:117]: Adjust for the audience:
```suggestion
      <p>No application was found at the provided URL.</p>
```

**Comment by miheer** [comment]
@Miciah PTAL

File: images/router/haproxy/conf/error-page-404.http [hunk 4]
```diff
@@ -0,0 +1,148 @@
+HTTP/1.0 404 Error
```
Comment (sgreene570) [inline images/router/haproxy/conf/error-page-404.http:1]: Is this new file identical to `images/router/haproxy/conf/error_code_pages/error-page-404.http`? We don't need both of them, right?

File: images/router/haproxy/conf/error_code_pages/error-page-503.http [hunk 5]
```diff
@@ -0,0 +1,140 @@
+HTTP/1.0 503 Service Unavailable
```
Comment (sgreene570) [inline images/router/haproxy/conf/error_code_pages/error-page-503.http:1]: If we are moving the 503 page to `images/router/haproxy/conf/error_code_pages/error-page-503.http`, we can delete `images/router/haproxy/conf/error-page-503.http`, right?

**Review by sgreene570 — COMMENTED** [review]
https://github.com/openshift/router/pull/140 added non-standard line endings to the 503 error page. We need to make sure that both of the error page files that we include use CRLF line terminators, otherwise we risk breaking customers. @miheer can you confirm that `error-page-404.http` is compliant with this (I'm assuming `error-page-503.http` was not modified by in this PR)?

File: images/router/haproxy/conf/error-page-404.http (hunk 4 above)
Comment (miheer) [reply images/router/haproxy/conf/error-page-404.http:1]: Yeah I had copied both files from the existing ones. Actually we need to have a backup somewhere so have not touched them.

File: images/router/haproxy/conf/error_code_pages/error-page-503.http (hunk 5 above)
Comment (miheer) [reply images/router/haproxy/conf/error_code_pages/error-page-503.http:1]: Yeah I had copied both files from the existing ones. So images/router/haproxy/conf/error_code_pages/ is for custom error pages and images/router/haproxy/conf/ for original pages. If custom pages is not provided then original pages will be served.

**Comment by miheer** [comment]
@Miciah PTAL. Can you please approve this after review ?

File: images/router/haproxy/conf/error_code_pages/error-page-503.http (hunk 5 above)
Comment (sgreene570) [reply images/router/haproxy/conf/error_code_pages/error-page-503.http:1]: If custom error pages are user provided than why do we need to copy the original error pages?

Comment (sgreene570) [reply images/router/haproxy/conf/error_code_pages/error-page-503.http:1]: If custom error pages are user provided than why do we need to copy the original error pages?

File: images/router/haproxy/reload-haproxy [hunk 6]
```diff
@@ -33,7 +33,7 @@ function haproxyHealthCheck() {
     while true; do
       local statusline=$(echo $'PROXY UNKNOWN\r\nGET / HTTP/1.1\r\n' | socat tcp-connect:localhost:${port} stdio | head -1)
 
-      if [[ "$statusline" == *" 503 "* ]]; then
+      if [ "$httpcode" == "503" || "$httpcode" == "404"]; then
```
Comment (candita) [inline images/router/haproxy/reload-haproxy:36]: Should this check $statusline?

File: images/router/haproxy/reload-haproxy [hunk 7]
```diff
@@ -53,7 +53,7 @@ function haproxyHealthCheck() {
   while true; do
     local httpcode=$(curl $timeout_opts -s -o /dev/null -I -H "Host: " -w "%{http_code}" ${url})
 
-    if [ "$httpcode" == "503" ]; then
+    if [[ "$statusline" == *" 503 "* || "$statusline" == *" 404 "* ]]; then
```
Comment (candita) [inline images/router/haproxy/reload-haproxy:56]: Should this check $httpcode?

File: images/router/haproxy/conf/haproxy-config.template [hunk 8]
```diff
@@ -131,7 +131,8 @@ defaults
   # backend service or alternatively, you can send a custom 503 error.
   #
   # server openshift_backend 127.0.0.1:8080
-  errorfile 503 /var/lib/haproxy/conf/error-page-503.http
+  errorfile 503 {{ env "ERRORFILE_503" "/var/lib/haproxy/conf/error-page-503.http" }}
```
Comment (candita) [inline images/router/haproxy/conf/haproxy-config.template:134]: I suggest naming the environment variable ROUTER_ERRORFILE_503/404, as seems to be a standard for the other environment variables to be prefixed with ROUTER

File: images/router/haproxy/conf/haproxy-config.template [hunk 9]
```diff
@@ -131,7 +131,8 @@ defaults
   # backend service or alternatively, you can send a custom 503 error.
```
Comment (candita) [inline images/router/haproxy/conf/haproxy-config.template:131]: Please update the comment here regarding custom error pages.  Is it still correct that one can uncomment the line?

File: images/router/haproxy/conf/haproxy-config.template [hunk 10]
```diff
@@ -434,6 +435,7 @@ backend openshift_default
   option forwardfor
   #option http-keep-alive
   option http-pretend-keepalive
+  http-request deny deny_status 404
```
Comment (candita) [inline images/router/haproxy/conf/haproxy-config.template:438]: This doesn't belong here any more, right?

Comment (miheer) [reply images/router/haproxy/conf/haproxy-config.template:438]: I did not get what you mean ? It main thing of this PR :)

Comment (miheer) [reply images/router/haproxy/conf/haproxy-config.template:438]: It is required to trigger 404 as haproxy generates 503.

**Comment by miheer** [comment]
@sgreene570 @Miciah PTAL

File: images/router/haproxy/reload-haproxy [hunk 11]
```diff
@@ -33,7 +33,7 @@ function haproxyHealthCheck() {
     while true; do
       local statusline=$(echo $'PROXY UNKNOWN\r\nGET / HTTP/1.1\r\n' | socat tcp-connect:localhost:${port} stdio | head -1)
 
-      if [[ "$statusline" == *" 503 "* ]]; then
+      if [ "$statusline" == "503" || "$statusline" == "404"]; then
```
Comment (Miciah) [inline images/router/haproxy/reload-haproxy:36]: You're missing a space before the closing bracket.  Also, please use double-braces (Bash-style, not POSIX style).  
```suggestion
      if [[ "$statusline" = "503" || "$statusline" = "404" ]]; then
```

File: images/router/haproxy/reload-haproxy [hunk 12]
```diff
@@ -53,7 +53,7 @@ function haproxyHealthCheck() {
   while true; do
     local httpcode=$(curl $timeout_opts -s -o /dev/null -I -H "Host: " -w "%{http_code}" ${url})
 
-    if [ "$httpcode" == "503" ]; then
+    if [[ "$httpcode" == *" 503 "* || "$httpcode" == *" 404 "* ]]; then
```
Comment (Miciah) [inline images/router/haproxy/reload-haproxy:56]: Why do you need globs?
```suggestion
    if [[ "$httpcode" = 503 || "$httpcode" = 404 ]]; then
```

**Comment by Miciah** [comment]
Do you need to add `    mkdir -p /var/lib/haproxy/conf/error_code_pages && \` in `Dockerfile.rhel8`?

**Comment by sgreene570** [comment]
> Do you need to add ` mkdir -p /var/lib/haproxy/conf/error_code_pages && \` in `Dockerfile.rhel8`?

I think the latest changes place both the default 404 and 504 error code file in `/var/lib/haproxy/conf`, so I don't think this would be needed.

**Comment by sgreene570** [comment]
> Do you need to add ` mkdir -p /var/lib/haproxy/conf/error_code_pages && \` in `Dockerfile.rhel8`?

Nothing is using the `error_code_pages` subdir in this PR's latest changes, right?

**Comment by miheer** [comment]
> mkdir -p /var/lib/haproxy/conf/error_code_pages && \

I don't think so? configmap will create a sub dir

**Comment by Miciah** [comment]
> > mkdir -p /var/lib/haproxy/conf/error_code_pages && \
> 
> I don't think so? configmap will create a sub dir

I thought the directory needed to exist to be used as a mount point.  Did the mount work without the directory's existence in your testing?

**Comment by Miciah** [comment]
> > Do you need to add ` mkdir -p /var/lib/haproxy/conf/error_code_pages && \` in `Dockerfile.rhel8`?
> 
> Nothing is using the `error_code_pages` subdir in this PR's latest changes, right?

No, but the intent is that the ingress operator mounts a configmap into that location and then points to the files in that directory using the environment variables.

**Comment by miheer** [comment]
> > > Do you need to add ` mkdir -p /var/lib/haproxy/conf/error_code_pages && \` in `Dockerfile.rhel8`?
> > 
> > 
> > Nothing is using the `error_code_pages` subdir in this PR's latest changes, right?
> 
> No, but the intent is that the ingress operator mounts a configmap into that location and then points to the files in that directory using the environment variables.

ACK added the changes

**Comment by sgreene570** [comment]
> ACK added the changes

Might as well air on the side of caution

/lgtm

File: images/router/haproxy/reload-haproxy (hunk 11 above)
Comment (candita) [reply images/router/haproxy/reload-haproxy]: Here, I think you need the `== *" 503 "*` format

File: images/router/haproxy/reload-haproxy (hunk 12 above)
Comment (candita) [reply images/router/haproxy/reload-haproxy]: Still `==` not `=`, right?

File: images/router/haproxy/conf/haproxy-config.template (hunk 10 above)
Comment (candita) [reply images/router/haproxy/conf/haproxy-config.template:438]: Ok, I misunderstood.

File: images/router/haproxy/reload-haproxy (hunk 11 above)
Comment (Miciah) [reply images/router/haproxy/reload-haproxy]: Ahh, this one does need globbing.  Somehow the `*` got transferred from here (where it is needed) to the check below (where it is not).  The `echo ... | socat ... | head -1` command will get something like "HTTP/1.0 404 Error", so we need to use globbing (`*" 404 "*`).

Comment (Miciah) [reply images/router/haproxy/reload-haproxy]: Below, where `curl ... -w "%{http_code}"` is used, the result will be just the code: either "404" or "503", so a simple string equality check suffices there.

File: images/router/haproxy/reload-haproxy (hunk 12 above)
Comment (Miciah) [reply images/router/haproxy/reload-haproxy]: `=` is traditional POSIX, but Bash allows `==` as well.  My preference is `=`, but `==` should be fine if there is a strong preference for it.

**Comment by sgreene570** [comment]
@miheer have you tested this PR against the router e2e tests? Do we need to make changes in origin to accommodate this PR?

**Comment by miheer** [comment]
> @miheer have you tested this PR against the router e2e tests? Do we need to make changes in origin to accommodate this PR?

I think we discussed on slack and now have pushed the necessary changes

**Comment by Miciah** [comment]
failed to acquire lease: resources not found

/test e2e-upgrade

**Comment by Miciah** [comment]
fail [github.com/onsi/ginkgo@v4.7.0-origin.0+incompatible/internal/leafnodes/runner.go:113]: Apr  8 17:48:02.018: Unexpected alerts fired or pending after the test run:
    
    alert ClusterOperatorDegraded fired for 16 seconds with labels: {condition="Degraded", endpoint="metrics", instance="10.0.175.10:9099", job="cluster-version-operator", name="authentication", namespace="openshift-cluster-version", pod="cluster-version-operator-554c8c6646-6wmfm", reason="WellKnownReadyController_SyncError", service="cluster-version-operator", severity="critical"}

**Comment by Miciah** [comment]
Many failures, mostly in sig-storage but also in sig-instrumentation, sig-cli, sig-auth, sig-auth, and sig-node.  Nothing that looks related to the router.

/test e2e-agnostic

**Comment by Miciah** [comment]
Don't care about e2e-metal-ipi-ovn-ipv6.
/skip

    [k8s.io] [sig-node] Pods Extended [k8s.io] Pod Container Status should never report success for a pending container [Suite:openshift/conformance/parallel] [Suite:k8s]

That isn't us.

/test e2e-agnostic

    [sig-instrumentation][Late] Alerts shouldn't exceed the 500 series limit of total series sent via telemetry from each cluster [Skipped:Disconnected] [Suite:openshift/conformance/parallel]

That isn't us.

/test e2e-aws-serial

    [sig-arch][Feature:ClusterUpgrade] Cluster should remain functional during upgrade [Disruptive] [Serial]
    disruption_tests: [sig-arch] Check if alerts are firing during or after upgrade success
    disruption_tests: [sig-storage] [sig-api-machinery] secret-upgrade
    disruption_tests: [sig-storage] [sig-api-machinery] configmap-upgrade

Not use.

/test e2e-upgrade

**Comment by Miciah** [comment]
The hold was for the globbing issue (https://github.com/openshift/router/pull/274#discussion_r609104513), so cancel that.

/hold cancel

**Comment by Miciah** [comment]
ns/openshift-monitoring pod/thanos-querier-86956cc6b9-hblmq node/ci-op-t4x24mwm-9c2de-ffhd2-worker-b-qd6cc - never deleted - reason/FailedCreatePodSandBox Failed to create pod sandbox: rpc error: code = Unknown desc = failed to create pod network sandbox k8s_thanos-querier-86956cc6b9-hblmq_openshift-monitoring_cb0cc4f7-ca87-458a-821b-fd085d1137e5_0(aa43f38c6c1d919713719a03fd74596654bd641e126a860a214b134fe9138b24): Multus: [openshift-monitoring/thanos-querier-86956cc6b9-hblmq]: error getting pod: Get "https://[api-int.ci-op-t4x24mwm-9c2de.origin-ci-int-gce.dev.openshift.com]:6443/api/v1/namespaces/openshift-monitoring/pods/thanos-querier-86956cc6b9-hblmq?timeout=1m0s": dial tcp 10.0.0.2:6443: i/o timeout

/test e2e-upgrade

**Pull request merged**